src/Graph.php

```diff
@@ -1,184 +1,172 @@
+use Microsoft\Graph\Core\NationalCloud;
+use Microsoft\Graph\Exception\ClientInitialisationException;
+use Microsoft\Graph\Http\GraphCollectionRequest;
+use Microsoft\Graph\Http\GraphRequest;
+use Microsoft\Graph\Http\HttpClientFactory;
+use Microsoft\Graph\Http\HttpClientInterface;
+
+/**
+ * Class Graph
+ *
+ * @category Library
+ * @package  Microsoft.Graph
+ * @license  https://opensource.org/licenses/MIT MIT License
+ * @link     https://graph.microsoft.io/
+ */
+class Graph
+{
+    /**
+    * The access_token provided after authenticating
+    * with Microsoft Graph (required)
+    *
+    * @var string
+    */
+    private $_accessToken;
+
+    /**
+    * The api version to use ("v1.0", "beta")
+    * Default is "v1.0"
+    *
+    * @var string
+    */
+    private $_apiVersion = GraphConstants::API_VERSION;
+
+    /**
+     * Host to use as the base URL and for authentication
+     * @var string
+     */
+    private $_nationalCloud = NationalCloud::GLOBAL;
+    
+    /**
+     * HttpClient to use for requests
+     * @var HttpClientInterface
+     */
+    private $_httpClient;
+
+
+    // TODO: Set middleware
+
+    /**
+     * Graph constructor.
+     *
+     * Creates a Graph client object used to make requests to the Graph API
+     * @param string $apiVersion
+     * @param string $nationalCloud
+     * @throws \InvalidArgumentException|Exception\ClientInitialisationException
+     */
+    public function __construct(string $apiVersion = GraphConstants::API_VERSION,
+                                string $nationalCloud = NationalCloud::GLOBAL)
+    {
+        if (!$apiVersion) {
+            throw new \InvalidArgumentException("Api version string cannot be empty");
+        }
+        if (!in_array($nationalCloud, NationalCloud::getValues())) {
+            throw new \InvalidArgumentException("Invalid national cloud passed. See NationalCloud constants");
+        }
+        $this->_apiVersion = $apiVersion;
+        $this->_nationalCloud = $nationalCloud;
+        $this->_httpClient = (new HttpClientFactory())->nationalCloud($nationalCloud)->createAdapter();
+    }
+
+    /**
+     * Sets HTTP client to any implementation of HttpClientInterface
+     *
+     * @param HttpClientInterface $httpClient
+     */
+    public function setHttpClient(HttpClientInterface $httpClient) {
+        $this->_httpClient = $httpClient;
+    }
+
+    /**
+     * Sets HTTP client using Guzzle request options
+     * Creates a Guzzle client and wraps it under an implementation of HttpClientInterface
+     *
+     * @param array $guzzleConfig
+     * @throws ClientInitialisationException
+     */
+    public function setHttpClientFromConfig(array $guzzleConfig) {
+        $this->_httpClient = (new HttpClientFactory())->clientConfig($guzzleConfig)
+                                                    ->nationalCloud($this->_nationalCloud)
+                                                    ->createAdapter();
+    }

```

Comment:
The $httpClient should only be set once. Ideally, this would happen in the Graph constructor. We want to prevent the customer from instantiating Graph object at time0 with a default httpClient, make some request, and then setting httpClient at a later time. If a customer needs a different httpClient, they can instantiate a new Graph object.

Reply:
Makese sense. Added in latest changes.